File: packages/relief_contracts/python/relief_contracts/forecast_response.py
"""ForecastResponseV1 (Section 12). Owner: Plan One + Plan Two. Version: 1.0.0.

The single response shape returned by every forecast provider (mock,
deterministic, relieffm). Python twin of ``src/forecast_response.ts``.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from .shared import DailySummaryEntryV1, ReasonFactorV1, TrajectoryPointV1, validate_contract_version


class ForecastProviderName(str, Enum):
    mock = "mock"
    deterministic = "deterministic"
    relieffm = "relieffm"


class DistressProbabilities(BaseModel):
    model_config = ConfigDict(extra="forbid")

    negative_balance: float = Field(ge=0, le=1)
    essential_reserve_violation: float = Field(ge=0, le=1)
    missed_obligation: float = Field(ge=0, le=1)


class ModelMetadataV1(BaseModel):
    """Populated only when provider == relieffm; deterministic sets this to None."""

    model_config = ConfigDict(extra="forbid")

    model_version: str
    calibration_version: Optional[str]
    inference_latency_ms: float = Field(ge=0)


class ForecastResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_version: str
    request_id: str
    forecast_id: str
    provider: ForecastProviderName
    provider_version: str
    generated_at: datetime
    valid_until: datetime
    confidence: float = Field(ge=0, le=1)
    is_stale: bool
    warnings: list[str]
    daily_summary: list[DailySummaryEntryV1]
    trajectories: list[TrajectoryPointV1]
    distress_probabilities: DistressProbabilities
    reason_factors: list[ReasonFactorV1]
    model_metadata: Optional[ModelMetadataV1]

    @field_validator("contract_version")
    @classmethod
    def _validate_contract_version(cls, v: str) -> str:
        return validate_contract_version(v)

    @model_validator(mode="after")
    def _model_metadata_only_for_relieffm(self) -> "ForecastResponseV1":
        if self.provider != ForecastProviderName.relieffm and self.model_metadata is not None:
            raise ValueError("model_metadata must be null unless provider is relieffm")
        return self

File: packages/relief_contracts/src/forecast_response.ts
import { z } from "zod";
import { ContractVersion } from "./shared";
import { DailySummaryEntryV1, TrajectoryPointV1, ReasonFactorV1 } from "./shared";

export const ForecastProviderName = z.enum(["mock", "deterministic", "relieffm"]);
export type ForecastProviderName = z.infer<typeof ForecastProviderName>;

/**
 * ModelMetadataV1 — populated only by the relieffm provider (Section 12).
 * The deterministic provider always sets `model_metadata: null`.
 */
export const ModelMetadataV1 = z.object({
  model_version: z.string(),
  calibration_version: z.string().nullable(),
  inference_latency_ms: z.number().nonnegative(),
});
export type ModelMetadataV1 = z.infer<typeof ModelMetadataV1>;

/**
 * ForecastResponseV1 (Section 12)
 *
 * Owner: Plan One + Plan Two (jointly)
 * Version: 1.0.0
 * Purpose: the single response shape returned by every forecast provider.
 *   No page, table, optimizer, or workflow may know which provider produced
 *   a given response — `provider` and `model_metadata` are the only signals,
 *   and both are optional to read.
 *
 * Required: contract_version, request_id, forecast_id, provider,
 *   provider_version, generated_at, valid_until, confidence, is_stale,
 *   warnings, daily_summary, trajectories, distress_probabilities,
 *   reason_factors, model_metadata (nullable — required key, nullable value).
 * Optional: none at 1.0.0.
 *
 * Validation rules:
 *   - confidence is 0..1.
 *   - model_metadata must be null for provider != "relieffm".
 *   - Every trajectory must reconcile: starting_balance_cents + inflow_cents
 *     - outflow_cents == ending_balance_cents (Section 45, Pass Ten).
 *
 * Migration notes: none yet (first version).
 */
export const ForecastResponseV1 = z
  .object({
    contract_version: ContractVersion,
    request_id: z.string(),
    forecast_id: z.string(),
    provider: ForecastProviderName,
    provider_version: z.string(),
    generated_at: z.string().datetime({ offset: true }),
    valid_until: z.string().datetime({ offset: true }),
    confidence: z.number().min(0).max(1),
    is_stale: z.boolean(),
    warnings: z.array(z.string()),
    daily_summary: z.array(DailySummaryEntryV1),
    trajectories: z.array(TrajectoryPointV1),
    distress_probabilities: z.object({
      negative_balance: z.number().min(0).max(1),
      essential_reserve_violation: z.number().min(0).max(1),
      missed_obligation: z.number().min(0).max(1),
    }),
    reason_factors: z.array(ReasonFactorV1),
    model_metadata: ModelMetadataV1.nullable(),
  })
  .refine((val) => val.provider === "relieffm" || val.model_metadata === null, {
    message: "model_metadata must be null unless provider is relieffm",
    path: ["model_metadata"],
  });
export type ForecastResponseV1 = z.infer<typeof ForecastResponseV1>;
